ddprof-lib/src/main/cpp/otel_process_ctx.cpp

```diff
@@ -0,0 +1,510 @@
+    data_out->telemetry_sdk_version = NULL;
+    data_out->telemetry_sdk_name = NULL;
+    data_out->resources = NULL;
+
+    // Allocate resources array with space for all pairs as a simplification (2 entries per pair + 1 for NULL terminator)
+    data_out->resources = (char **) calloc(count * 2 + 1, sizeof(char *));
+    if (!data_out->resources) return false;
+
+    int resources_index = 0;
+
+    // Decode each key-value pair
+    for (int i = 0; i < count; i++) {
+      // Check str 16 header for key (0xda)
+      if ((unsigned char)*ptr++ != 0xda) return false;
+
+      // Read key length (2 bytes, big endian)
+      uint16_t key_len = ((uint8_t)*ptr << 8) | (uint8_t)*(ptr + 1);
+      ptr += 2;
+
+      // Get pointer to key (not null-terminated)
+      char *key_not_terminated = ptr;
+      ptr += key_len;
+
+      // Check str 16 header for value (0xda)
+      if ((unsigned char)*ptr++ != 0xda) return false;
+
+      // Read value length (2 bytes, big endian)
+      uint16_t value_len = ((uint8_t)*ptr << 8) | (uint8_t)*(ptr + 1);
+      ptr += 2;
+
+      // Read value
+      char *value = (char *) calloc(value_len + 1, 1);
+      if (!value) return false;
+      memcpy(value, ptr, value_len);
+      value[value_len] = '\0';
+      ptr += value_len;
+
+      // Assign to appropriate field based on key
+      if (key_len == strlen("deployment.environment.name") && memcmp(key_not_terminated, "deployment.environment.name", strlen("deployment.environment.name")) == 0) {
+        data_out->deployment_environment_name = value;
+      } else if (key_len == strlen("host.name") && memcmp(key_not_terminated, "host.name", strlen("host.name")) == 0) {
+        data_out->host_name = value;
+      } else if (key_len == strlen("service.instance.id") && memcmp(key_not_terminated, "service.instance.id", strlen("service.instance.id")) == 0) {
+        data_out->service_instance_id = value;
+      } else if (key_len == strlen("service.name") && memcmp(key_not_terminated, "service.name", strlen("service.name")) == 0) {
+        data_out->service_name = value;
+      } else if (key_len == strlen("service.version") && memcmp(key_not_terminated, "service.version", strlen("service.version")) == 0) {
+        data_out->service_version = value;
+      } else if (key_len == strlen("telemetry.sdk.language") && memcmp(key_not_terminated, "telemetry.sdk.language", strlen("telemetry.sdk.language")) == 0) {
+        data_out->telemetry_sdk_language = value;
+      } else if (key_len == strlen("telemetry.sdk.version") && memcmp(key_not_terminated, "telemetry.sdk.version", strlen("telemetry.sdk.version")) == 0) {
+        data_out->telemetry_sdk_version = value;
+      } else if (key_len == strlen("telemetry.sdk.name") && memcmp(key_not_terminated, "telemetry.sdk.name", strlen("telemetry.sdk.name")) == 0) {
+        data_out->telemetry_sdk_name = value;
+      } else {
+        // Unknown key, put it into resources
+        char *key = (char *) calloc(key_len + 1, 1);
```

Comment:
Could we avoid the manual allocs ? if we have to handle frees on error we will make mistakes 
example not freeing other resources (`data_out->resources` )

Reply:
Note that the resources are being cleared in the caller:

```c
    if (!otel_process_ctx_decode_payload(mapping->otel_process_payload, &data)) {
      otel_process_ctx_read_data_drop(data);
```

For avoiding manual allocs, any suggestions? Were you thinking something closer to -- allocate a big block of and then just memcpy into it?

(Also the reading support is mostly only for debugging/tests and can even be disabled individually, as it's not particularly interesting in actual production apps)

Reply:
Disclaimer: when I initially wrote this I forgot that we had to be compatible with C :smile: 
In C++ the types would have defined the ownership.

In C, we can try to be more symmetrical on alloc / free code paths: having clear init / free functions to look for.
I'm not sure we need to polish this right now.